File: src/main/java/ru/zdoher/japs/repositories/aggregation/RepositoryAggregationImpl.java
package ru.zdoher.japs.repositories.aggregation;

import com.netflix.hystrix.contrib.javanica.annotation.HystrixCommand;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.data.mongodb.core.aggregation.Aggregation;
import org.springframework.data.mongodb.core.aggregation.SampleOperation;

public class RepositoryAggregationImpl<T> implements RepositoryAggregation<T> {
    private static final int RANDOM_COUNT = 1;

    private MongoTemplate mongoTemplate;

    public RepositoryAggregationImpl(MongoTemplate mongoTemplate) {
        this.mongoTemplate = mongoTemplate;
    }

    @Override
    public T getRandom(Class<T> tClass) {

        SampleOperation match = Aggregation.sample(RANDOM_COUNT);
        Aggregation aggregation = Aggregation.newAggregation(match);

        return mongoTemplate.aggregate(aggregation, tClass, tClass).getUniqueMappedResult();
    }
}

File: src/main/java/ru/zdoher/japs/repositories/aggregation/RepositoryAggregationImpl.kt
package ru.zdoher.japs.repositories.aggregation

import org.springframework.data.mongodb.core.ReactiveMongoTemplate
import org.springframework.data.mongodb.core.aggregation.Aggregation
import reactor.core.publisher.Mono

class RepositoryAggregationImpl<T>(private val mongoTemplate: ReactiveMongoTemplate) : RepositoryAggregation<T> {

    override fun getRandom(tClass: Class<T>): Mono<T> {
        val match = Aggregation.sample(RANDOM_COUNT.toLong())
        val aggregation = Aggregation.newAggregation(match)
        return mongoTemplate.aggregate(aggregation, tClass, tClass).last()
    }

    companion object {
        private const val RANDOM_COUNT = 1
    }

}